man/man1/admin.1
.\"	@(#)admin.1	6.3 (Berkeley) 10/21/04
.tr ~
.tr $%
.if t .tr #\fB\(em\fP
.if n .tr #-
.if n .ds D " -- 
.if t .ds D _
.tr @|
.nr f 0
.bd S B 3
.de SP
.if n .ul
\%[\fB\(em\\$1\fR\\c
.if n .ul 0
\\$2\\$3
..
.de SF
.if n .ul
\%[\fB\(em\\$1\fR]
.if n .ul 0
..
.de ZZ
.hc ^
\fB^...\fR
.hc
..
.de AR
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\fB\(em\\$1\\fR
\\$2 \\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de C1
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\\$1
\\$2 \\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de A1
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\fB\(em\\$1\fR[\fI\\$2\fR]
\\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de A2
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 5
.TP 15
\fB\(em\\$1\fI\\$2\fR
\\$3 \\$4 \\$5 \\$6 \\$7 \\$8 \\$9
.nr f 1
..
.de FI
.PP
.TP 30
\\$1
\\$2
.i0
..
.ds F)  \fB\s-1FILES\s+1\fR
.ds W)  \fB\s-1WARNINGS\s+1\fR
.ds X)  \fB\s-1EXAMPLES\s+1\fR
.ds T)  \fB\s-1TABLE\s+1\fR
.ds K)  \fB\s-1DATA KEYWORDS\s+1\fR
.ds D)  \fB\s-1DDDDD\s+1\fR
.ds M)  \fB\s-1MR\s+1\fR
.ds R)  \fB\s-1RELEASE NUMBER\s+1\fR
.ds S)  \fB\s-1SCCS\s+1\fR
.ds I)  \fB\s-1SID\s+1\fR
.de F1
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 13
.TP 7
\fB\\$1\fI\\$2\fR
\\$3
.nr f 1
..
.de F2
.if \\nf \{ \
.    RE
.    nr f 0 \}
.PP
.RS 13
.TP 7
\fB\\$1[\fI\\$2\fR]
\\$3
.nr f 1
..
.if n .ds )Q '
.if n .ds )G `
.if t .ds )Q \\(aa
.if t .ds )G \\(ga
.if t .ds )S \\|
.TH ADMIN 1 "October 21, 2004"
.SH NAME
admin \- create and administer \s-1SCCS\s0 files
.SH SYNOPSIS
.na
.B admin
.SF n
.SP i \%[name] ]
.SP r rel ]
.SP t \%[name] ]
.if n .br
.SP f flag\%[flag-val] ]
.ZZ
.SP d flag\%[flag-val] ]
.ZZ
.br
.SP a login ]
.ZZ
.SP e login ]
.ZZ
.SP m \%[mrlist] ]
.SP y \%[comment] ]
.SF h
.SF z
file
.ZZ
.ad
.SH DESCRIPTION
.I Admin
is used to create new \*(S) files
and change parameters of existing ones.
Arguments to
.I admin\c
,
which may appear in any order,
consist of keyletter arguments,
which begin with \fB\(em,\fR
and named files
(note that \*(S) file names must begin with the characters
\fBs.\fR)\fB.\fR
If a named file doesn't exist,
it is created,
and its parameters are initialized according to
the specified keyletter arguments.
Parameters not initialized by a keyletter argument
are assigned a default value.
If a named file does exist,
parameters corresponding to specified keyletter arguments
are changed,
and other parameters are left as is.
.PP
If a directory is named,
.I admin
behaves as though each file in the directory were
specified as a named file,
except that non-\*(S) files
(last component of the pathname does not begin with \fBs.\fR)
and unreadable files
are silently ignored.
If a name of \fB\(em\fR is given, the standard input is read;
each line of the standard input is taken to be the name of an \*(S) file
to be processed.
Again, non-\*(S) files and unreadable files are silently ignored.
.PP
The keyletter arguments are as follows.
Each is explained as though only one named file is to be processed
since the effects of the arguments
apply independently to
each named file.
.AR n This
keyletter indicates that a new \*(S) file is to be created.
.A1 i name The
\fIname\fR of a file from which the text for a new \*(S) file is to be taken.
The text constitutes the first delta of the file
(see
.B \(emr
keyletter for delta numbering scheme).
If the
.B i
keyletter is used,
but the file name is omitted,
the text is obtained by reading the standard input
until an end-of-file is encountered.
If this keyletter is omitted,
then the \*(S) file is created empty.
Only one \*(S) file may be created
by an
.I admin
command on which the
.B i
keyletter is supplied.
Using a single
.I admin
to create two or more \*(S) files require that they be created
empty
(no
.B \(emi
keyletter).
Note that the
.B \(emi
keyletter implies the
.B \(emn
keyletter.
.A2 r rel The
.I rel\c
ease into which the initial delta is inserted.
This keyletter may be used only if the
.B \(emi
keyletter is also used.
If the
.B \(emr
keyletter is not used,
the initial delta is inserted into release 1.
The level of the initial delta is always 1
(by default initial deltas are named 1.1).
.A1 t name The
\fIname\fR of a file from which descriptive text for the \*(S) file
is to be taken.
If the
.B \(emt
keyletter is used
and
.I admin
is creating a new \*(S) file
(the
.B \(emn
and/or
.B \(emi
keyletters also used), the descriptive text file name
must also
be supplied.
In the case of existing \*(S) files:
(1) a
.B \(emt
keyletter without a file name causes removal of descriptive text
(if any) currently in the \*(S) file, and (2) a
.B \(emt
keyletter with a file name causes text (if any) in the
named file to replace
the descriptive text (if any) currently in the \*(S) file.
.A2 f flag This
keyletter specifies a \fIflag\fR, and, possibly, a value for the \fIflag\fR, to
be placed in the \*(S) file.
Several
.B f
keyletters may be supplied on a single
.I admin
command line.
The allowable \fIflag\fRs and their values are:
.F1 b ~ Allows
use of
the
.B \(emb
keyletter on a
.IR get\^ (1)
command to create branch deltas.
.F1 c ceil The
highest release
(i.e., ``ceiling''),
a number less than or equal to 9999, which may be
retrieved by a
.IR get\^ (1)
command for editing.
The default value for an unspecified
.B c
flag is 9999.
.F1 f floor The
lowest release
(i.e., ``floor''),
a number greater than 0 but less than 9999,
which may be retrieved by a
.IR get\^ (1)
command for editing.
The default value for an unspecified
.B f
flag is 1.
.F1 d \s-1SID\s0 The
default delta number
(\*(I)) to be used by a
.IR get\^ (1)
command.
.F1 i ~ Causes
the "No id keywords (ge6)" message issued by
.IR get\^ (1)
or
.IR delta\^ (1)
to be treated as a fatal error.
In the absence of this flag, the message is only a warning.
The message is issued if no
\*(S) identification keywords
(see
.IR get\^ (1))
are found
in the text retrieved or stored in the \*(S) file.
.F1 j ~ Allows
concurrent
.IR get\^ (1)
commands for editing
on the same \*(I) of an \*(S) file.
This allows multiple concurrent updates to the same
version of the \*(S) file.
.F1 l list A
\fIlist\fR of releases to which deltas can no longer be made
.I (get
.B \(eme
against one of these ``locked'' releases fails).
The \fIlist\fR has the following syntax:
.F1 ~ ~ <list>
::= <range> \(or <list> , <range>
.br
<range>~::=	\*(R) \(or \fBa\fR
.F1 ~ ~ The
character \fBa\fR in the \fIlist\fR is equivalent to specifying
.I "all releases"
for the named \*(S) file.
.F1 n ~ Causes
.IR delta\^ (1)
to create a ``null'' delta
in each of those releases (if any) being skipped when a delta is made
in a
.I new
release (e.g., in making delta 5.1 after delta 2.7, releases 3 and 4
are skipped).
These null deltas serve as ``anchor points'' so that branch deltas
may later be created from them.
The absence of this flag causes skipped releases to be
non-existent in the \*(S) file
preventing branch deltas from being created from them in the future.
.F1 q text User
definable text substituted for all occurrences
of the $Q$ keyword in \*(S) file text retrieved by
.IR get\^ (1).
.F1 m mod \fIMod\fRule
name of the \*(S) file
substituted for all occurrences of the $M$ keyword in \*(S) file
text retrieved by
.IR get\^ (1).
If the
.B m
flag is not specified, the value
assigned is the name of the \*(S) file with the leading
\fBs.\fR removed.
.F1 t type \fIType\fR
of module in the \*(S) file substituted for all occurrences of
$Y$ keyword in \*(S) file text retrieved by
.IR get\^ (1).
.F2 v pgm Causes
.IR delta\^ (1)
to prompt for Modification Request (\*(M)) numbers as the
reason for creating a delta.
The optional value specifies the name of an \*(M) number validity
checking program
(see
.IR delta\^ (1)).
.A2 d flag Causes
removal (deletion) of the specified \fIflag\fR from an \*(S) file.
The
.B \(emd
keyletter
may be specified only when processing existing
\*(S) files.
Several
.B \(emd
keyletters may be supplied on a single
.I admin
command.
See the
.B \(emf
keyletter for allowable \fIflag\fR names.
.F1 l list A
\fIlist\fR of releases to be ``unlocked''.
See the
.B \(emf
keyletter for a description of the
.B l
flag and the syntax of a \fIlist\fR.
.A2 a login A
\fIlogin\fR name, or numerical \fB\s-2UNIX\s+2\fR group \fB\s-2ID\s+2\fR,
to be added to the list of users which
may make deltas (changes) to the \*(S) file.
A group \fB\s-2ID\s+2\fR is equivalent to specifying all
\fIlogin\fR names common to that group \fB\s-2ID\s+2\fR.
Several
.B a
keyletters may be used
on a single
.I admin
command line.
As many \fIlogin\fRs, or numerical group \fB\s-2ID\s+2\fRs, as desired may be on the list simultaneously.
If the list of users is empty,
then anyone may add deltas.
.A2 e login A
\fIlogin\fR name, or numerical group \fB\s-2ID\s+2\fR, to be erased from the list of users
allowed to make deltas (changes) to the \*(S) file.
Specifying a group \fB\s-2ID\s+2\fR is equivalent to specifying all
\fIlogin\fR names common to that group \fB\s-2ID\s+2\fR.
Several
.B e
keyletters may be used on a single
.I admin
command line.
.A1 y comment The
.I comment
text is inserted
into the \*(S) file as a comment
for the initial delta in a manner identical to that
of
.IR delta\^ (1).
Omission of the \fB\(emy\fR keyletter results in a default
comment line being inserted in the form:
.sp
.if n \{\
.br
date and time created YY/MM/DD HH:MM:SS by
<logname>\}
.ie t \{\
.br
~~~~~date and time created \fB\s-2YY/MM/DD HH:MM:SS\s+2\fR by <logname>\}
.sp
The
.B \(emy
keyletter
is valid only if the
.B \(emi
and/or
.B \(emn
keyletters are
specified (i. e. a new \*(S) file is being created).
.A1 m mrlist The
list of Modification Requests (\*(M)) numbers is inserted into the \*(S) file
as the reason for creating the initial delta in a manner
identical to
.IR delta\^ (1).
The
.B v
flag must be set and the \*(M) numbers are validated if the
.B v
flag has a value (the name of an \*(M) number validation program).
Diagnostics will occur if the
.B v
flag is not set or \*(M) validation fails.
.AR h Causes
.I admin
to check the structure of the \*(S) file
(see
.I sccsfile\c
(5)), and to compare a newly
computed check-sum (the sum of all the characters in the \*(S) file
except those in the first line) with the check-sum that is stored
in the first line of the \*(S) file.
Appropriate error diagnostics are produced.
.C1 ~ This
keyletter inhibits writing on the file, so that it
nullifies the effect of any other keyletters supplied, and
is, therefore, only meaningful when processing existing files.
.AR z The
\*(S) file check-sum is recomputed and stored in the first line
of the \*(S) file
(see
.B\(emh,
above).
.C1 ~ Note
that use of this keyletter on a truly corrupted file
may prevent future detection of the corruption.
.i0
.SH FILES
The last component of
all \*(S) file names must be of the form ``\fBs.\fP\fIfilename\fP''.
New \*(S) files are given mode 444
(see
.IR chmod\^ (1)).
Write permission
in the pertinent directory is,
of course,
required to create a file.
All writing done by
.I admin
is to a temporary x-file,
called ``\fBx.\fP\fIfilename\fP'',
(see
.IR get\^ (1)),
created with mode 444 if the
.I admin
command is creating a new \*(S) file, or with the same mode
as the \*(S) file if it exists.
After successful execution of
.I admin\c
,
the \*(S) file is removed (if it exists), and the x-file
is renamed with the name of the \*(S) file.
This ensures that changes are made to the \*(S) file only
if no errors occurred.
.PP
It is recommended that directories containing \*(S) files be mode
755
and that \*(S) files themselves be mode 444.
The mode of the directories allows
only the owner to modify
\*(S) files contained in the directories.
The mode of the \*(S) files prevents any modification
at all except by \*(S) commands.
.PP
If it should be necessary to patch an \*(S) file for any reason,
the mode may be changed to 644 by the owner
allowing use of
.IR ed\^ (1).
\fICare must be taken!.\fR
The edited file should
.I always
be processed by an
.I admin
.B \(emh
to check for corruption followed by an
.I admin
.B \(emz
to generate a proper check-sum.
Another
.I admin
.B \(emh
is recommended to ensure the \*(S) file is valid.
.PP
.I Admin
also makes use of
a transient lock file
(called ``\fBz.\fP\fIfilename\fP''),
which is used to prevent simultaneous updates to the \*(S) file by different users.
See
.IR get\^ (1)
for further information.
.SH "SEE ALSO"
.na
get(1),
delta(1),
prs(1),
what(1),
sccshelp(1),
ed(1),
sccsfile(5)
.br
.I "Source Code Control System User's Guide"
by L. E. Bonanni and C. A. Salemi.
.ad
.SH DIAGNOSTICS
Use
.IR sccshelp\^ (1)
for explanations.
.tr ~~
.tr $$
.tr @@


usr.bin/sccscmds/cmd/admin.c
# include "../hdr/defines.h"
# include "../hdr/had.h"

static char Sccsid[] = "@(#)admin.c	4.4	1/3/99";

/*
	Program to create new SCCS files and change parameters
	of existing ones. Arguments to the program may appear in
	any order and consist of keyletters, which begin with '-',
	and named files. Named files which do not exist are created
	and their parameters are initialized according to the given
	keyletter arguments, or are given default values if the
	corresponding keyletters were not supplied. Named files which
	do exist have those parameters corresponding to given key-letter
	arguments changed and other parameters are left as is.

	If a directory is given as an argument, each SCCS file within
	the directory is processed as if it had been specifically named.
	If a name of '-' is given, the standard input is read for a list
	of names of SCCS files to be processed.
	Non-SCCS files are ignored.

	Files created are given mode 444.
*/

# define MINR 1		/* minimum release number */
# define MAXR 9999	/* maximum release number */
# define MAXNAMES 9
# define COPY 0
# define NOCOPY 1

char *ifile, *tfile;
char *z;	/* for validation program name */
char had[26], had_flag[26], rm_flag[26];
char	*Comments, *Mrs;
char Valpgm[] = "/usr/bin/val";
int irel, fexists, num_files;
int	VFLAG = 0;
int	Domrs;
char *Sflags[];
char *anames[MAXNAMES], *enames[MAXNAMES];
char *flag_p[26];
int asub, esub;
int check_id;
int Did_id;

main(argc,argv)
int argc;
char *argv[];
{
	register int j;
	register char *p;
	char c, f;
	int i, testklt;
	extern admin();
	extern int Fcnt;
	struct sid sid;

	/*
	Set flags for 'fatal' to issue message, call clean-up
	routine and terminate processing.
	*/
	Fflags = FTLMSG | FTLCLN | FTLEXIT;

	testklt = 1;

	/*
	The following loop processes keyletters and arguments.
	Note that these are processed only once for each
	invocation of 'main'.
	*/
	for(j=1; j<argc; j++)
		if(argv[j][0] == '-' && (c = argv[j][1])) {
			p = &argv[j][2];
			switch (c) {

			case 'i':	/* name of file of body */
				ifile = p;
				break;

			case 't':	/* name of file of descriptive text */
				tfile = p;
				break;
			case 'm':	/* mr flag */
				Mrs = p;
				break;
			case 'y':	/* comments flag for entry */
				Comments = p;
				break;

			case 'd':	/* flags to be deleted */
				testklt = 0;
				if (!(f = *p))
					fatal("d has no argument (ad1)");
				p = &argv[j][3];

				switch (f) {

				case IDFLAG:	/* see 'f' keyletter */
				case BRCHFLAG:	/* for meanings of flags */
				case VALFLAG:
				case TYPEFLAG:
				case MODFLAG:
				case NULLFLAG:
				case FLORFLAG:
				case CEILFLAG:
				case DEFTFLAG:
					if (*p) {
						sprintf(Error, "value after %c flag (ad12)",f);
						fatal(Error);
					}
					break;

				default:
					fatal("unknown flag (ad3)");
				}

				if (rm_flag[f - 'a']++)
					fatal("flag twice (ad4)");
				break;

			case 'f':	/* flags to be added */
				testklt = 0;
				if (!(f = *p))
					fatal("f has no argument (ad5)");
				p = &argv[j][3];

				switch (f) {

				case IDFLAG:	/* id-kwd message (err/warn) */
				case BRCHFLAG:	/* branch */
				case NULLFLAG:	/* null deltas */
					if (*p) {
						sprintf(Error, "value after %c flag (ad13)",f);
						fatal(Error);
					}
					break;

				case VALFLAG:	/* mr validation */
					VFLAG++;
					if (*p)
						z = p;
					break;

				case FLORFLAG:	/* floor */
					if ((i = patoi(p)) == -1)
						fatal("floor not numeric (ad22)");
					if ((size(p) > 5) || (i < MINR) ||
							(i > MAXR))
						fatal("floor out of range (ad23)");
					break;

				case CEILFLAG:	/* ceiling */
					if ((i = patoi(p)) == -1)
						fatal("ceiling not numeric (ad24)");
					if ((size(p) > 5) || (i < MINR) ||
							(i > MAXR))
						fatal("ceiling out of range (ad25)");
					break;

				case DEFTFLAG:	/* default sid */
					if (!(*p))
						fatal("no default sid (ad14)");
					chksid(sid_ab(p,&sid),&sid);
					break;

				case TYPEFLAG:	/* type */
				case MODFLAG:	/* module name */
					if (!(*p)) {
						sprintf(Error, "flag %c has no value (ad2)",f);
						fatal(Error);
					}
					break;

				default:
					fatal("unknown flag (ad3)");
				}

				if (had_flag[f - 'a']++)
					fatal("flag twice (ad4)");
				flag_p[f - 'a'] = p;
				break;

			case 'r':	/* initial release number supplied */
				if ((irel = patoi(p)) == -1)
					fatal("r arg not numeric (ad6)");
				if ((size(p) > 5) || (irel < MINR) ||
						(irel > MAXR))
					fatal("r out of range (ad7)");
				break;

			case 'n':	/* creating new SCCS file */
			case 'h':	/* only check hash of file */
			case 'z':	/* zero the input hash */
				break;

			case 'a':	/* user-name allowed to make deltas */
				testklt = 0;
				if (!(*p))
					fatal("bad a argument (ad8)");
				if (asub > MAXNAMES)
					fatal("too many 'a' keyletters (ad9)");
				anames[asub++] = p;
				break;

			case 'e':	/* user-name to be removed */
				testklt = 0;
				if (!(*p))
					fatal("bad e argument (ad10)");
				if (esub > MAXNAMES)
					fatal("too many 'e' keyletters (ad11)");
				enames[esub++] = p;
				break;

			default:
				fatal("unknown key letter (cm1)");
			}

			if (had[c - 'a']++ && testklt++)
				fatal("key letter twice (cm2)");
			argv[j] = 0;
		}
		else
			num_files++;

	if (num_files == 0)
		fatal("missing file arg (cm3)");

	if (HADI && num_files > 1) /* only one file allowed with `i' */
		fatal("more than one file (ad15)");

	setsig();

	/*
	Change flags for 'fatal' so that it will return to this
	routine (main) instead of terminating processing.
	*/
	Fflags &= ~FTLEXIT;
	Fflags |= FTLJMP;

	/*
	Call 'admin' routine for each file argument.
	*/
	for (j=1; j<argc; j++)
		if (p = argv[j])
			do_file(p,admin);

	exit(Fcnt ? 1 : 0);
}


/*
	Routine that actually does admin's work on SCCS files.
	Existing s-files are copied, with changes being made, to a
	temporary file (x-file). The name of the x-file is the same as the
	name of the s-file, with the 's.' replaced by 'x.'.
	s-files which are to be created are processed in a similar
	manner, except that a dummy s-file is first created with
	mode 444.
	At end of processing, the x-file is renamed with the name of s-file
	and the old s-file is removed.
*/

struct packet gpkt;	/* see file defines.h */
char	Zhold[BUFSIZ];	/* temporary z-file name */

USXALLOC();		/* defines alloc() and free() */

admin(afile)
char *afile;
{
	struct deltab dt;	/* see file defines.h */
	struct stats stats;	/* see file defines.h */
	FILE *iptr;
	register int k;
	register char *cp, *q;
	char command[80];
	char line[512];
	int i;			/* used in forking procedure */
	int status;
	extern nfiles;
	extern had_dir;

	if (setjmp(Fjmp))	/* set up to return here from 'fatal' */
		return;		/* and return to caller of admin */

	if (HADI && had_dir) /* directory not allowed with `i' keyletter */
		fatal("directory named with `i' keyletter (ad26)");

	fexists = exists(afile);

	if (HADI)
		HADN = 1;
	if (HADI || HADN) {
		if (HADM && !VFLAG)
			fatal("MRs not allowed (de8)");

		if (VFLAG && !HADM)
			fatal("MRs required (de10)");

	}

	if (!HADI && HADR)
		fatal("r only allowed with i (ad16)");

	if (HADN && HADT && !(*tfile))
		fatal("t has no argument (ad17)");

	if (HADN && HADD)
		fatal("d not allowed with n (ad18)");

	if (HADN && fexists) {
		sprintf(Error,"file %s exists (ad19)",afile);
		fatal(Error);
	}

	if (!HADN && !fexists) {
		sprintf(Error,"file %s does not exist (ad20)",afile);
		fatal(Error);
	}
	/*
	   Check for '-h' flag.  If set, create child process and
	   invoke 'get' to examine format of SCCS file.
	*/

	if (HADH) {
		/*
		   fork here so 'admin' can execute 'val' to
		   check for a corrupted file.
		*/
		if ((i = fork()) < 0)
			fatal("cannot fork, try again");
		if (i == 0) {		/* child */
			/*
			   perform 'val' with appropriate keyletters
			*/
			sprintf(command, "/usr/bin/val -s %s", afile);
			execl("/bin/sh","/bin/sh","-c", command, 0);
			sprintf(Error,"cannot execute '%s'",Valpgm);
			fatal(Error);
		}
		else {
			wait(&status);	   /* wait on status from 'execl' */
			if (status)
				fatal("corrupted file (co6)");
			return;		/* return to caller of 'admin' */
		}
	}

	/*
	Lock out any other user who may be trying to process
	the same file.
	*/
	if (!HADH && lockit(copy(auxf(afile,'z'),Zhold),2,getpid()))
		fatal("cannot create lock file (cm4)");

	if (fexists)
		sinit(&gpkt,afile,1);	/* init pkt & open s-file */
	else {
		xfcreat(afile,0444);	/* create dummy s-file */
		sinit(&gpkt,afile,0);	/* and init pkt */
	}

	if (!HADH)
		/*
		   set the flag for 'putline' routine to open
		   the 'x-file' and allow writing on it.
		*/
		gpkt.p_upd = 1;

	if (HADZ) {
		gpkt.do_chksum = 0;	/* ignore checksum processing */
		gpkt.p_ihash = 0;
	}

	/*
	Get statistics of latest delta in old file.
	*/
	if (!HADN) {
		stats_ab(&gpkt,&stats);
		gpkt.p_wrttn++;
		newstats(&gpkt,line,"0");
	}

	if (HADN) {		/*   N E W   F I L E   */

		/*
		Beginning of SCCS file.
		*/
		sprintf(line,"%c%c%s\n",CTLCHAR,HEAD,"00000");
		putline(&gpkt,line);

		/*
		Statistics.
		*/
		newstats(&gpkt,line,"0");

		dt.d_type = 'D';	/* type of delta */

		/*
		Set initial release, level, branch and
		sequence values.
		*/
		if (HADR)
			dt.d_sid.s_rel = irel;
		else
			dt.d_sid.s_rel = 1;
		dt.d_sid.s_lev = 1;
		dt.d_sid.s_br = dt.d_sid.s_seq = 0;

		time(&dt.d_datetime);		/* get time and date */

		copy(logname(),dt.d_pgmr);	/* get user's name */

		dt.d_serial = 1;
		dt.d_pred = 0;

		del_ba(&dt,line);	/* form and write */
		putline(&gpkt,line);	/* delta-table entry */

		/*
		If -m flag, enter MR numbers
		*/

		if (Mrs) {
			mrfixup();
			if (z && valmrs(&gpkt,z))
				fatal("invalid MRs (de9)");
			putmrs(&gpkt);
		}

		/*
		Enter comment line for `chghist'
		*/

		if (HADY) {
			sprintf(line,"%c%c ",CTLCHAR,COMMENTS);
			putline(&gpkt,line);
			putline(&gpkt,Comments);
			putline(&gpkt,"\n");
		}
		else {
			/*
			insert date/time and pgmr into comment line
			*/
			cmt_ba(&dt,line);
			putline(&gpkt,line);
		}
		/*
		End of delta-table.
		*/
		sprintf(line,CTLSTR,CTLCHAR,EDELTAB);
		putline(&gpkt,line);

		/*
		Beginning of user-name section.
		*/
		sprintf(line,CTLSTR,CTLCHAR,BUSERNAM);
		putline(&gpkt,line);
	}
	else
		/*
		For old file, copy to x-file until user-name section
		is found.
		*/
		flushto(&gpkt,BUSERNAM,COPY);

	/*
	Write user-names to be added to list of those
	allowed to make deltas.
	*/
	if (HADA)
		for (k = 0; k < asub; k++) {
			sprintf(line,"%s\n",anames[k]);
			putline(&gpkt,line);
		}

	/*
	Do not copy those user-names which are to be erased.
	*/
	if (HADE && !HADN)
		while ((cp = getline(&gpkt)) &&
				!(*cp++ == CTLCHAR && *cp == EUSERNAM)) {
			for (k = 0; k < esub; k++) {
				cp = &gpkt.p_line;
				while (*cp)	/* find and */
					cp++;	/* zero newline */
				*--cp = '\0';	/* character */

				if (equal(enames[k],&gpkt.p_line)) {
					/*
					Tell getline not to output
					previously read line.
					*/
					gpkt.p_wrttn = 1;
					break;
				}
				else
					*cp = '\n';	/* restore newline */
			}
		}

	if (HADN) {		/*   N E W  F I L E   */

		/*
		End of user-name section.
		*/
		sprintf(line,CTLSTR,CTLCHAR,EUSERNAM);
		putline(&gpkt,line);
	}
	else
		/*
		For old file, copy to x-file until end of
		user-names section is found.
		*/
		if (!HADE)
			flushto(&gpkt,EUSERNAM,COPY);

	/*
	For old file, read flags and their values (if any), and
	store them. Check to see if the flag read is one that
	should be deleted.
	*/
	if (!HADN)
		while ((cp = getline(&gpkt)) &&
				(*cp++ == CTLCHAR && *cp == FLAG)) {

			gpkt.p_wrttn = 1;	/* don't write previous line */

			cp += 2;	/* point to flag character */
			k = *cp - 'a';

			if (!had_flag[k] && !rm_flag[k]) {
				had_flag[k] = 2;	/* indicate flag is */
							/* from file, not */
							/* from arg list */

				if (*++cp != '\n') {	/* get flag value */
					q = alloc(size(gpkt.p_line)-5);
					copy(++cp,q);
					flag_p[k] = q;
					while (*q)	/* find and */
						q++;	/* zero newline */
					*--q = '\0';	/* character */
				}
			}
			else
				if (rm_flag[k])
					had_flag[k] = 0;
		}


	/*
	Write out flags.
	*/
	for (k = 0; k < 26; k++)
		if (had_flag[k]) {
			if (flag_p[k])
				sprintf(line,"%c%c %c %s\n",
					CTLCHAR,FLAG,'a'+k,flag_p[k]);
			else
				sprintf(line,"%c%c %c\n",
					CTLCHAR,FLAG,'a'+k);

			putline(&gpkt,line);

			if (had_flag[k] == 2) {	/* flag was taken from file */
				had_flag[k] = 0;
				if (flag_p[k]) {
					free(flag_p[k]);
					flag_p[k] = 0;
				}
			}
		}

	if (HADN) {
		/*
		Beginning of descriptive (user) text.
		*/
		sprintf(line,CTLSTR,CTLCHAR,BUSERTXT);
		putline(&gpkt,line);
	}
	else
		/*
		Write out BUSERTXT record which was read in
		above loop that processes flags.
		*/
		gpkt.p_wrttn = 0;
		putline(&gpkt,0);

	/*
	Get user description, copy to x-file.
	*/
	if (HADT) {
		if (*tfile) {
			iptr = xfopen(tfile,0);
			fgetchk(line,512,iptr,tfile,&gpkt);
			fclose(iptr);
		}

		/*
		If old file, ignore any previously supplied
		commentary. (i.e., don't copy it to x-file.)
		*/
		if (!HADN)
			flushto(&gpkt,EUSERTXT,NOCOPY);
	}

	if (HADN) {		/*   N E W  F I L E   */

		/*
		End of user description.
		*/
		sprintf(line,CTLSTR,CTLCHAR,EUSERTXT);
		putline(&gpkt,line);

		/*
		Beginning of body (text) of first delta.
		*/
		sprintf(line,"%c%c %u\n",CTLCHAR,INS,1);
		putline(&gpkt,line);

		if (HADI) {		/* get body */

			/*
			Set indicator to check lines of body of file for
			keyword definitions.
			If no keywords are found, a warning
			will be produced.
			*/
			check_id = 1;
			/*
			Set indicator that tells whether there
			were any keywords to 'no'.
			*/
			Did_id = 0;
			if (*ifile)
				iptr = xfopen(ifile,0);	/* from a file */
			else
				iptr = stdin;	/* from standard input */

			/*
			Read and copy to x-file, while checking
			first character of each line to see that it
			is not the control character (octal 1).
			Also, count lines read, and set statistics'
			structure appropriately.
			The 'fgetchk' routine will check for keywords.
			*/
			stats.s_ins = fgetchk(line,512,iptr,ifile,&gpkt);
			stats.s_del = stats.s_unc = 0;

			/*
			If no keywords were found, issue warning.
			*/
			if (!Did_id) {
				if (had_flag[IDFLAG - 'a'])
					fatal("no id keywords (cm6)");
				else
					fprintf(stderr,"%s\n","No id keywords (cm7)");
			}

			check_id = 0;
			Did_id = 0;
		}

		/*
		End of body of first delta.
		*/
		sprintf(line,"%c%c %u\n",CTLCHAR,END,1);
		putline(&gpkt,line);
	}
	else {
		/*
		Indicate that EOF at this point is ok, and
		flush rest of (old) s-file to x-file.
		*/
		gpkt.p_chkeof = 1;
		while (getline(&gpkt)) ;
	}

	/*
	Flush the buffer, take care of rewinding to insert
	checksum and statistics in file, and close.
	*/
	flushline(&gpkt,&stats);

	/*
	Change x-file name to s-file, and delete old file.
	Unlock file before returning.
	*/
	if (!HADH) {
		rename(auxf(&gpkt,'x'),&gpkt);
		xrm(&gpkt);
		unlockit(auxf(afile,'z'),getpid());
	}
}


fgetchk(strp,len,inptr,file,pkt)
register char *strp;
register int len;
FILE *inptr;
register char *file;
register struct packet *pkt;
{
	register int k;

	for (k = 1; fgets(strp,len,inptr); k++) {
		if (*strp == CTLCHAR) {
			sprintf(Error,"%s illegal data on line %d (ad21)",
				file,k);
			fatal(Error);
		}

		if (check_id)
			chkid(strp);

		putline(pkt,strp);
	}
	return(k - 1);
}


clean_up()
{
	xrm(&gpkt);
	if (!HADH)
		unlockit(Zhold,getpid());
	if (HADN)
		unlink(&gpkt);
}


cmt_ba(dt,str)
register struct deltab *dt;
char *str;
{
	register char *p;

	p = str;
	*p++ = CTLCHAR;
	*p++ = COMMENTS;
	*p++ = ' ';
	copy("date and time created",p);
	while (*p++)
		;
	--p;
	*p++ = ' ';
	date_ba(&dt->d_datetime,p);
	while (*p++)
		;
	--p;
	*p++ = ' ';
	copy("by",p);
	while (*p++)
		;
	--p;
	*p++ = ' ';
	copy(dt->d_pgmr,p);
	while (*p++)
		;
	--p;
	*p++ = '\n';
	*p = 0;
	return(str);
}


putmrs(pkt)
struct packet *pkt;
{
	register char **argv;
	char str[64];
	extern char *Varg[];

	for (argv = &Varg[VSTART]; *argv; argv++)
		sprintf(str,"%c%c %s\n",CTLCHAR,MRNUM,*argv);
		putline(pkt,str);
}
